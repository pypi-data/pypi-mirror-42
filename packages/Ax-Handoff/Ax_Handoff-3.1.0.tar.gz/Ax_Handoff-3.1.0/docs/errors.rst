==============================================================================
Errors & Exceptions
==============================================================================

The ``axonchisel.handoff.error`` module contains all Exceptions raised by 
this package:
    
- ``SerializeError``:
    Error encoding an object to wire format.
        
- ``UnserializeError``
    Error decoding an object from wire format.

- ``DataTamperedError``
    Subclass of UnserializeError indicating data
    was tampered with or corrupted in transit.

