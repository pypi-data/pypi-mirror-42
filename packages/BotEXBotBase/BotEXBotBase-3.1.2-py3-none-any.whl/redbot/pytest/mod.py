import pytest

__all__ = ["mod"]


@pytest.fixture
def mod(config, monkeypatch):
    from redbot.core import Config

    with monkeypatch.context() as m:
        m.setattr(Config, "get_conf", lambda *args, **kwargs: config)
        from redbot.core import modlog

        modlog._register_defaults()
        return modlog
