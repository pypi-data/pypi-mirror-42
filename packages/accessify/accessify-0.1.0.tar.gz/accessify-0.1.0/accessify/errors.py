"""
Provide errors for accessibility levels.
"""
INACCESSIBLE_DUE_TO_ITS_PROTECTION_LEVEL_EXCEPTION_MESSAGE = \
    '{class_name}.{method_name}() is inaccessible due to its protection level'


class InaccessibleDueToItsProtectionLevelException(Exception):
    """
    Inaccessible due to its protection level exception.
    """

    def __init__(self, message):
        self.message = message