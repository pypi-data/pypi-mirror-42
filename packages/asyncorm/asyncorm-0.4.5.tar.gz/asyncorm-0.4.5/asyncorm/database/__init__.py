from asyncorm.database.db_manager import PostgresManager
from asyncorm.database.db_cursor import Cursor

__all__ = ["PostgresManager", "Cursor"]
