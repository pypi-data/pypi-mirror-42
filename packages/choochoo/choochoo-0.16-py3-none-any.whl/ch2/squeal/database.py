
from contextlib import contextmanager

from sqlalchemy import create_engine, event
from sqlalchemy.engine import Engine
from sqlalchemy.orm import sessionmaker
from sqlalchemy.sql.functions import count

from . import *
from .support import Base
from ..command.args import DATABASE, NamespaceWithVariables, NO_OP, parser
from ..lib.log import make_log

# mention these so they are "created"
Source,  Interval,
ActivityGroup, ActivityJournal, ActivityTimespan,
Topic, TopicJournal,
StatisticName, StatisticJournal, StatisticJournalInteger, StatisticJournalFloat, StatisticJournalText,
Segment, SegmentJournal
Pipeline
MonitorJournal
Constant, SystemConstant
ActivitySimilarity, ActivityNearby


# https://stackoverflow.com/questions/13712381/how-to-turn-on-pragma-foreign-keys-on-in-sqlalchemy-migration-script-or-conf
@event.listens_for(Engine, "connect")
def fk_pragma_on_connect(dbapi_con, _con_record):
    cursor = dbapi_con.cursor()
    cursor.execute("PRAGMA foreign_keys=ON;")  # https://www.sqlite.org/pragma.html#pragma_foreign_keys
    cursor.execute("PRAGMA temp_store=MEMORY;")  # https://www.sqlite.org/pragma.html#pragma_temp_store
    cursor.execute("PRAGMA threads=4;")  # https://www.sqlite.org/pragma.html#pragma_threads
    cursor.execute("PRAGMA cache_size=-1000000;")  # 1GB  https://www.sqlite.org/pragma.html#pragma_cache_size
    cursor.execute("PRAGMA secure_delete=OFF;")  # https://www.sqlite.org/pragma.html#pragma_secure_delete
    # cursor.execute("PRAGMA journal_mode=WAL;")  # https://www.sqlite.org/wal.html
    cursor.close()


@event.listens_for(Engine, 'close')
def analyxe_pragma_on_close(dbapi_con, _con_record):
    cursor = dbapi_con.cursor()
    cursor.execute("PRAGMA optimize;")  # https://www.sqlite.org/pragma.html#pragma_optimize
    cursor.close()


class Database:

    def __init__(self, args, log):
        self._log = log
        path = args.file(DATABASE)
        self._log.info('Using database at %s' % path)
        self.engine = create_engine('sqlite:///%s' % path, echo=False)
        self.session = sessionmaker(bind=self.engine)
        self.__create_tables()

    def __create_tables(self):
        if self.is_empty(tables=True):
            self._log.info('Creating tables')
            Base.metadata.create_all(self.engine)

    @contextmanager
    def session_context(self):
        session = self.session()
        try:
            yield session
            session.commit()
        except:
            session.rollback()
            raise
        finally:
            session.close()

    def is_empty(self, tables=False):
        if tables:
            # https://stackoverflow.com/questions/33053241/sqlalchemy-if-table-does-not-exist
            return not self.engine.dialect.has_table(self.engine, Source.__tablename__)
        else:
            with self.session_context() as s:
                n_topics = s.query(count(Topic.id)).scalar()
                n_activities = s.query(count(ActivityGroup.id)).scalar()
                n_statistics = s.query(count(StatisticName.id)).scalar()
            return not (n_topics + n_activities + n_statistics)


def add(s, instance):
    s.add(instance)
    return instance


def connect(args):
    '''
    Bootstrap from commandline-like args.
    '''
    p = parser()
    if len(args) == 1:
        args = args[0].split()
    elif args:
        args = list(args)
    else:
        args = []
    args.append(NO_OP)
    ns = NamespaceWithVariables(p.parse_args(args))
    log = make_log(ns)
    db = Database(ns, log)
    return ns, log, db

