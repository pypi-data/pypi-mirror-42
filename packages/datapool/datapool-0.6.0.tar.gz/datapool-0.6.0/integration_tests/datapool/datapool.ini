[db]
connection_string = sqlite+pysqlite:////Users/uweschmitt/Projects/eawag_datapool/src/integration_tests/lz/.simple.db

[backup_landing_zone]
folder = 

[r]
extension = .r
executable = /usr/local/bin/R

[matlab]
extension = .m
executable = /Applications/MATLAB_R2016a.app/bin/matlab

[julia]
extension = .jl
version = 0.5.2
executable = /usr/local/bin/julia

[python]
extension = .py
executable = /Users/uweschmitt/Projects/eawag_datapool/venv/bin/python3

[logging]
config_file = ./logging_config.yaml

[log_receiver]
port = 5559

[worker]
port = 5555
count = 5

[server]
pid_file = /Users/uweschmitt/.datapool/datapool.pid

[landing_zone]
folder = /Users/uweschmitt/Projects/eawag_datapool/src/integration_tests/lz

