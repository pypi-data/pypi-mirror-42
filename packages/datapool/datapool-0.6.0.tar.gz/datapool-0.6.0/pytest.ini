[pytest]
addopts = -rs --full-trace
testpaths = tests
