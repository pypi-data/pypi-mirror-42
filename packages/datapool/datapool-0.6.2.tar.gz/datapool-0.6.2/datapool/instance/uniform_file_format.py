# encoding: utf-8
from __future__ import print_function, division, absolute_import

import csv
import datetime
import io
import os

from datapool.utils import iter_to_list


def _check_header(header):

    header = list(header)

    if "site" in header:
        header.remove("site")
    else:
        for n in "xyz":
            if n not in header:
                raise ValueError(
                    "header {!r} must have either 'site' or 'x', 'y' and 'z'".format(
                        header
                    )
                )
            header.remove(n)

    header = set(header)
    fix_column_names = set(("timestamp", "parameter", "value"))
    if header != fix_column_names:
        missing = ["'{}'".format(name) for name in fix_column_names - header]
        invalid = ["'{}'".format(name) for name in header - fix_column_names]
        msg = ""
        if missing:
            msg += "missing: {}".format(", ".join(sorted(missing)))
        if invalid:
            msg += "invalid: {}".format(", ".join(sorted(invalid)))
        raise ValueError("invalid header ({})".format(msg))


@iter_to_list
def read_from_file(path):
    if not os.path.exists(path):
        raise ValueError("{} does not exist".format(path))

    with open(path, "r", encoding="ascii") as fh:
        yield from _read_from_fh(fh)


@iter_to_list
def read_from_string(data):
    fh = io.StringIO(data)
    yield from _read_from_fh(fh)


def can_be_converted_to_float(value):
    try:
        value = float(value)
    except ValueError:
        return False
    return True


def parse_timestamp(timestamp):
    return datetime.datetime.strptime(timestamp, "%Y-%m-%d %H:%M:%S")


def _read_from_fh(fh):
    reader = csv.reader(fh, delimiter=";")
    header = [f.strip() for f in next(reader)]
    _check_header(header)
    for i, row in enumerate(reader):
        if len(row) < len(header):
            # i + 1, we skipped header
            raise ValueError("row {} ({!r}) is incomplete".format(i + 1, row))
        if len(row) > len(header):
            # i + 1, we skipped header
            raise ValueError("row {} ({!r}) is to long".format(i + 1, row))

        row = (cell.strip() for cell in row)
        row_dict = dict(zip(header, row))
        yield row_dict


@iter_to_list
def check_rows(row_dicts):

    if not row_dicts:
        return

    for i, row_dict in enumerate(row_dicts):

        value = row_dict["value"]
        if not can_be_converted_to_float(value):
            yield "row {}: value '{}' is not a float".format(i, value)
        else:
            row_dict["value"] = float(value)

        timestamp = row_dict["timestamp"]
        try:
            row_dict["timestamp"] = parse_timestamp(timestamp)
        except ValueError:
            yield "row {}: timstamp '{}' either has wrong format or is invalid".format(
                i, timestamp
            )

        has_xyz = all(name in row_dict and row_dict[name] is not None for name in "xyz")
        not_empty = ("parameter",)
        if not has_xyz:
            not_empty += ("site",)

        for name in not_empty:
            if row_dict[name].strip() == "":
                yield "row {}: field {} is empty".format(i, name)


@iter_to_list
def to_signals(row_dicts):

    for row in row_dicts:
        for f in "xyz":
            if f in row:
                row["coord_" + f] = row[f]
                del row[f]

    from .domain_objects import Signal

    return list(map(Signal, row_dicts))
