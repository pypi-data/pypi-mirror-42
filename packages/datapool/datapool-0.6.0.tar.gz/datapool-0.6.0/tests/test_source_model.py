# encoding: utf-8
from __future__ import absolute_import, division, print_function

from functools import partial

import pytest

from datapool.database import dump_db, setup_db
from datapool.errors import ConsistencyError
from datapool.instance.parameters_model import \
    check_and_commit as check_and_commit_parameters
from datapool.instance.pretty_printers import pretty_print
from datapool.instance.source_model import check, check_and_commit
from datapool.instance.source_type_model import \
    check_and_commit as check_and_commit_st
from datapool.instance.yaml_parsers import (parse_parameters, parse_source,
                                            parse_source_type)

from .yaml_helpers import load

load_parameters = partial(load, parse_parameters)
load_source = partial(load, parse_source)
load_source_type = partial(load, parse_source_type)


def test_load_valid_source(data_path, regtest, regtest_print, regtest_logger, lz):
    with regtest_logger():
        excs, sources = load_source(lz)
    assert len(sources) == 4, excs
    regtest_print("source objects read from yamls:")
    for source in sources:
        pretty_print(source, file=regtest)
    assert not excs, str(excs)


def test_load_empty_source_yml(data_path, regtest_print, regtest_logger, lz):
    lz.overwrite_file(
        "data/FloDar/FloDar_python/source.yaml", data_path("yamls/empty.yaml")
    )
    lz.overwrite_file(
        "data/FloDar/FloDar_julia/source.yaml", data_path("yamls/empty.yaml")
    )
    lz.overwrite_file("data/FloDar/FloDar_R/source.yaml", data_path("yamls/empty.yaml"))
    lz.overwrite_file(
        "data/FloDar/FloDar_matlab/source.yaml", data_path("yamls/empty.yaml")
    )

    with regtest_logger():
        excs, sources = load_source(lz)

    assert not sources
    assert excs
    for exc in excs:
        regtest_print(exc)


def test_load_invalid_source_yml(data_path, regtest_print, regtest_logger, lz):
    lz.overwrite_file(
        "data/FloDar/FloDar_python/source.yaml", data_path("yamls/source_invalid.yaml")
    )

    with regtest_logger():
        excs, sources = load_source(lz)

    assert len(sources) == 3
    assert excs
    for exc in excs:
        regtest_print(exc)


@pytest.fixture(scope="function")
def engine_with_example_data(config, lz):
    __, (parameters,) = load_parameters(lz)
    engine = setup_db(config.db)
    check_and_commit_parameters(parameters, engine)

    excs, (source_type,) = load_source_type(lz)
    assert not excs

    check_and_commit_st(source_type, engine)
    return engine


def test_commit_valid_source(
    data_path,
    regtest_logger,
    regtest,
    regtest_print,
    config,
    lz,
    engine_with_example_data,
):

    with regtest_logger():
        excs, sources = load_source(lz)
        assert not excs

    source = min(sources, key=lambda s: s.name)

    regtest_print("source objects read from yaml:")
    pretty_print(source, file=regtest)

    with regtest_logger():
        for source_dbo in check_and_commit(source, engine_with_example_data):
            regtest_print()
            regtest_print("source row:", source_dbo)
            pretty_print(source_dbo, file=regtest)
            regtest_print()

    dump_db(config.db, file=regtest)


def test_commit_invalid_source_yml(
    data_path, regtest_print, regtest_logger, lz, config
):

    __, (parameters,) = load_parameters(lz)

    engine = setup_db(config.db)
    check_and_commit_parameters(parameters, engine)

    lz.overwrite_file(
        "data/FloDar/FloDar_python/source.yaml",
        data_path("yamls/source_unknown_parameters.yaml"),
    )

    excs, sources = load_source(lz)

    with regtest_logger():
        regtest_print("EXCEPTIONS:")
        for source in sources:
            for e in check(source, engine):
                regtest_print(e)


def test_update_modified_source_yml(
    data_path, regtest, regtest_logger, config, lz, engine_with_example_data
):
    def setup():
        excs, sources = load_source(lz)
        assert not excs, str(excs)

        source, = [s for s in sources if s.name == "flodar_python"]

        print("source object read from yaml:", file=regtest)
        pretty_print(source, file=regtest)
        print(file=regtest)

        with regtest_logger():
            for source_dbo in check_and_commit(source, engine_with_example_data):
                print("source row after update:", source_dbo, file=regtest)
                pretty_print(source_dbo, file=regtest)
                print(file=regtest)

    setup()

    lz.overwrite_file(
        "data/FloDar/FloDar_python/source.yaml",
        data_path("yamls/source_averaging_added.yaml"),
    )
    with regtest_logger():
        excs, sources = load_source(lz)
    assert not excs, str(excs)
    source, = [s for s in sources if s.name == "flodar_python"]

    print("source object read from yaml:", file=regtest)
    pretty_print(source, file=regtest)

    # modified source should work:
    with regtest_logger():
        for dbo in check_and_commit(source, engine_with_example_data):
            print(dbo, file=regtest)

    # load a modified (but syntactically valid) source.yaml
    lz.overwrite_file(
        "data/FloDar/FloDar_python/source.yaml", data_path("yamls/source_modified.yaml")
    )
    with regtest_logger():
        excs, sources = load_source(lz)
    assert not excs, str(excs)
    source, = [s for s in sources if s.name == "flodar_python"]

    print("source object read from yaml:", file=regtest)
    pretty_print(source, file=regtest)

    # modified source should not raise errors:
    for dbo in check_and_commit(source, engine_with_example_data):
        print(dbo, file=regtest)

    # invalid modification
    source.averaging[0].parameter = "was_Temperatur"

    with pytest.raises(ConsistencyError) as e:
        with regtest_logger():
            check_and_commit(source, engine_with_example_data)
        print(file=regtest)
        print(e.value, file=regtest)
        print(file=regtest)

    dump_db(config.db, file=regtest)
