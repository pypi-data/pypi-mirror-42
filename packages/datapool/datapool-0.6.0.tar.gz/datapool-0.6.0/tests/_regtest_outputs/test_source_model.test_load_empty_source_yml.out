INFO     eawag.datapool  use parse_source to parse all files from <tmpdir_from_fixture>/landing_zone
INFO     eawag.datapool  load <tmpdir_from_fixture>/landing_zone/data/FloDar/FloDar_matlab/source.yaml
WARNING  eawag.datapool  file <tmpdir_from_fixture>/landing_zone/data/FloDar/FloDar_matlab/source.yaml is empty
INFO     eawag.datapool  load <tmpdir_from_fixture>/landing_zone/data/FloDar/FloDar_R/source.yaml
WARNING  eawag.datapool  file <tmpdir_from_fixture>/landing_zone/data/FloDar/FloDar_R/source.yaml is empty
INFO     eawag.datapool  load <tmpdir_from_fixture>/landing_zone/data/FloDar/FloDar_python/source.yaml
WARNING  eawag.datapool  file <tmpdir_from_fixture>/landing_zone/data/FloDar/FloDar_python/source.yaml is empty
INFO     eawag.datapool  load <tmpdir_from_fixture>/landing_zone/data/FloDar/FloDar_julia/source.yaml
WARNING  eawag.datapool  file <tmpdir_from_fixture>/landing_zone/data/FloDar/FloDar_julia/source.yaml is empty
file <tmpdir_from_fixture>/landing_zone/data/FloDar/FloDar_matlab/source.yaml is empty
file <tmpdir_from_fixture>/landing_zone/data/FloDar/FloDar_R/source.yaml is empty
file <tmpdir_from_fixture>/landing_zone/data/FloDar/FloDar_python/source.yaml is empty
file <tmpdir_from_fixture>/landing_zone/data/FloDar/FloDar_julia/source.yaml is empty
